File: backend/src/models/person.py
from sqlalchemy import Column, Integer, String, Date, DateTime, ForeignKey, Table
from sqlalchemy.orm import relationship
from sqlalchemy.dialects.postgresql import JSON
from datetime import datetime
from ..database import Base

# Association table for person-to-person relationships
person_relationships = Table(
    'person_relationships',
    Base.metadata,
    Column('person_id', Integer, ForeignKey('people.id', ondelete='CASCADE'), primary_key=True),
    Column('related_person_id', Integer, ForeignKey('people.id', ondelete='CASCADE'), primary_key=True),
    Column('relationship_type', String(100)),  # e.g., "spouse", "child", "parent", "sibling", etc.
    Column('created_at', DateTime, default=datetime.utcnow),
    Column('updated_at', DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
)

class Person(Base):
    __tablename__ = 'people'

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(255), nullable=False, index=True)
    phone_number = Column(String(20), nullable=True)
    date_of_birth = Column(Date, nullable=True)
    ssn_encrypted = Column(String(255), nullable=True)  # Encrypted SSN
    email = Column(String(255), nullable=True, index=True)
    employer = Column(String(255), nullable=True)
    occupation = Column(String(255), nullable=True)
    mailing_address = Column(JSON, nullable=True)  # JSON object with address fields
    physical_address = Column(JSON, nullable=True)  # JSON object with address fields
    
    # Trustor-related fields
    trustor_is_living = Column(Integer, default=1, nullable=True)  # 1 = living, 0 = deceased
    trustor_death_certificate_received = Column(Integer, default=0, nullable=True)
    trustor_of_sound_mind = Column(Integer, default=1, nullable=True)
    trustor_has_relinquished = Column(Integer, default=0, nullable=True)
    trustor_relinquished_date = Column(Date, nullable=True)
    trustor_reling_doc_received = Column(Integer, default=0, nullable=True)
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    is_active = Column(Integer, default=1, nullable=False)
    
    # Self-referential many-to-many relationship
    relationships = relationship(
        'Person',
        secondary=person_relationships,
        primaryjoin=id == person_relationships.c.person_id,
        secondaryjoin=id == person_relationships.c.related_person_id,
        backref='related_to'
    )

    def __repr__(self):
        return f"<Person(id={self.id}, name='{self.name}', email='{self.email}')>"

File: go-backend/internal/models/person.go
package models

import (
	"encoding/json"
	"time"
)

type Person struct {
	ID                           int              `gorm:"primaryKey" json:"id"`
	Name                         string           `gorm:"size:255;index;not null" json:"name"`
	PhoneNumber                  *string          `gorm:"size:20" json:"phone_number"`
	DateOfBirth                  *string          `gorm:"type:date" json:"date_of_birth"`
	SSNEncrypted                 *string          `gorm:"size:255" json:"-"`
	Email                        *string          `gorm:"size:255;index" json:"email"`
	Employer                     *string          `gorm:"size:255" json:"employer"`
	Occupation                   *string          `gorm:"size:255" json:"occupation"`
	MailingAddress               *json.RawMessage `gorm:"type:jsonb" json:"mailing_address"`
	PhysicalAddress              *json.RawMessage `gorm:"type:jsonb" json:"physical_address"`
	TrustorIsLiving              *int             `gorm:"default:1" json:"trustor_is_living"`
	DateOfDeath                  *string          `gorm:"type:date" json:"date_of_death"`
	TrustorDeathCertReceived     *int             `gorm:"column:trustor_death_certificate_received;default:0" json:"trustor_death_certificate_received"`
	TrustorOfSoundMind           *int             `gorm:"default:1" json:"trustor_of_sound_mind"`
	TrustorHasRelinquished       *int             `gorm:"default:0" json:"trustor_has_relinquished"`
	TrustorRelinquishedDate      *string          `gorm:"type:date" json:"trustor_relinquished_date"`
	TrustorRelingDocReceived     *int             `gorm:"column:trustor_reling_doc_received;default:0" json:"trustor_reling_doc_received"`
	IsActive                     int              `gorm:"default:1;not null" json:"is_active"`
	CreatedAt                    time.Time        `gorm:"not null" json:"created_at"`
	UpdatedAt                    time.Time        `gorm:"not null" json:"updated_at"`

	HasSSN bool `gorm:"-" json:"has_ssn"`
}

func (Person) TableName() string { return "people" }

type PersonRelationshipRow struct {
	PersonID         int       `gorm:"primaryKey;column:person_id"`
	RelatedPersonID  int       `gorm:"primaryKey;column:related_person_id"`
	RelationshipType string    `gorm:"size:100;column:relationship_type"`
	CreatedAt        time.Time `gorm:"column:created_at"`
	UpdatedAt        time.Time `gorm:"column:updated_at"`
}

func (PersonRelationshipRow) TableName() string { return "person_relationships" }
